### PR TITLE
Add interface module

### PR DESCRIPTION
**Pull request by blackHatMonkey**
This patch will add interface module which will replace all the functions related to setting up the interfaces in the main program. It will also add a compatibility check for wireless adapters which will reduce the errors related to interface incompatibility. Also the automatic interface selection algorithm has been greatly improved.

<h3>Change log</h3>
 Compatibility check
<p>
Added a wireless adapter compatibility check where the adapter is now being checked for <i>AP</i> and <i>monitor</i> modes.
</p>
 Various errors
<p>
Added numerous errors for better debugging. 
</p>

### CODE REVIEW

File: wifiphisher/interfaces.py [hunk 1]
```diff
@@ -0,0 +1,440 @@
+"""
+This module was made to handle all the interface related operations for
+Wifiphisher.py
+"""
+
+import subprocess
+import os
+import re
+
+DEVNULL = open(os.devnull, "w")
```
Comment (sophron) [inline wifiphisher/interfaces.py]: You should import and use the DN from <a href="https://github.com/sophron/wifiphisher/blob/master/wifiphisher/constants.py#L17">constants.py</a> file.

File: wifiphisher/interfaces.py [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+                          available.
+
+    Dependencies:
+        word_in_sentence
+    """
+
+    # Initialize variable names
+    device_name = str()
+    has_monitor = bool()
+    has_ap = bool()
```
Comment (sophron) [inline wifiphisher/interfaces.py]: This is not needed in Python as there is no such thing as "variable initialization".

File: wifiphisher/interfaces.py [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+    # run the command to get all the wireless devices
+    device_command = subprocess.Popen(["iw", "dev"], stdout=subprocess.PIPE,
+                                      stderr=subprocess.PIPE)
```
Comment (sophron) [inline wifiphisher/interfaces.py]: I think we should follow a strict top-down modular design here.

As this module includes plenty of Unix commands the best way to do it is by wrapping each of these commands in a separate method. For example, `cmd_iwconfig` should execute `iwconfig` command based on the given parameters and return the result (or an error code).

All these wrappers should build the required command and then leverage a method that exists in a lower layer (according to the design), that could be `exec_cmd`. This method will execute the given command by calling `Popen` or `os.open` accordingly.

In other words, I'm asking for something similar to SQL prepared statements.

File: wifiphisher/interfaces.py [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+    info_command = subprocess.Popen(["iw", device_name, "info"],
+                                    stdout=subprocess.PIPE, stderr=DEVNULL)
+
+    # list of all the info
+    info = info_command.communicate()[0].split("\n")
+
+    # check info to see if monitor or AP mode is supported
+    for line in info:
+        # set has_monitor to True if monitor mode is supported
+        if line == "\t\t * monitor":
```
Comment (sophron) [inline wifiphisher/interfaces.py]: In all `line == SOMETHING`,  `SOMETHING` appears to be a constant. We want to define this in `constants.py` and import it from there.

File: wifiphisher/interfaces.py [hunk 5]
```diff
@@ -0,0 +1,440 @@
...
+                                  interface.
+
+        NoMonitorInterfaceFound : In case of not finding a valid monitor
+                                 interface.
+
+    Dependencies:
+        get_wireless_interfaces
+        check_compatibility
+        get_interface
+        select_interface_automatically
```
Comment (sophron) [inline wifiphisher/interfaces.py]: `Dependencies` here may be misleading. We use the word `dependency` in Python when we refer to a required module. I would change the name to `Used methods:`

File: wifiphisher/interfaces.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+        # find the interfaces automatically
+        monitor_interface, ap_interface =\
+                select_interface_automatically(wireless_interfaces)
+
+    return monitor_interface, ap_interface
+
+
+def select_interface_automatically(interfaces):
+    """
+    Return a tuple containing monitor interface fallowed by ap interface which
```
Comment (sophron) [inline wifiphisher/interfaces.py]: * followed by AP interface which both were selected automatically.

File: wifiphisher/pywifiphisher.py [hunk 7]
```diff
@@ -1006,49 +936,48 @@ def run():
...
+        print err
+        shutdown()
+    except interfaces.NoMonitorInterfaceFound as err:
+        print err
+        shutdown()
+
+    # display selected interfaces to the user
+    print "Selecting {0} interface for jamming.".format(mon_iface)
+    print "Selecting {0} interface for access point creation.".format(ap_iface)
+    print ""
```
Comment (sophron) [inline wifiphisher/pywifiphisher.py]: We should use the right prefixes when we output information.

File: wifiphisher/pywifiphisher.py [hunk 8]
```diff
@@ -1006,49 +936,48 @@ def run():
...
+        mon_iface, ap_iface = interfaces.get_interfaces(args.jamminginterface,
+                                                        args.apinterface)
+        # TODO: this line should be removed once all the wj_iface have been
+        # removed
+        wj_iface = mon_iface
+    except interfaces.NotEnoughInterfacesFound as err:
+        print err
+        print ("HINT: Wifiphisher requires two wireless adapters of which "
+               "at least one should support AP mode and one should support"
+               "monitor mode.")
```
Comment (sophron) [inline wifiphisher/pywifiphisher.py]: I would remove the 'HINT' as this is not really a hint.

**Comment by sophron** [comment]
This is an important PR that should significantly improve the stability of the tool. Do you think you can write some tests for it?

I've done some audit (need to do more) and probably tomorrow I'll test its functionality. It would also be great if you could list in more detail some of the features of this PR.

File: wifiphisher/interfaces.py (hunk 2 above)
Comment (blackHatMonkey) [inline wifiphisher/interfaces.py]: I understand that in python there is no variable initialization. It's more of a declaration rather than initialization and I placed them to make the code clearer.

File: wifiphisher/interfaces.py (hunk 4 above)
Comment (blackHatMonkey) [inline wifiphisher/interfaces.py]: I have to disagree with this one. If I understand correctly the idea behind ```constants.py``` is to reuse the constants that were previously in the main module. Why would ```SOMETHING``` be placed there? ```SOMETHING``` is only used in ```interfaces``` module and will not be used anywhere else in the tool.

File: wifiphisher/interfaces.py (hunk 5 above)
Comment (blackHatMonkey) [inline wifiphisher/interfaces.py]: I wrote this to serve as a warning to anyone who might change one of those functions since ```get_interfaces``` function is dependent on these functions. Maybe ```dependent functions``` would be a better name for it.

File: wifiphisher/pywifiphisher.py (hunk 8 above)
Comment (blackHatMonkey) [inline wifiphisher/pywifiphisher.py]: Did you mean to remove the whole print statement or just the ```HINT```?

File: wifiphisher/interfaces.py (hunk 3 above)
Comment (blackHatMonkey) [inline wifiphisher/interfaces.py]: I'm not sure I fully understand what you are describing. Is this what you mean?

Create  ```exec_cmd``` function. This function will run any command and return its value.

Create ```cmd_iwconfig``` function. This function uses ```exec_cmd``` to run a command and catches any errors.

Now anywhere in the code we can use ```cmd_iwconfig``` to run the ```iwconfig``` command.

**Comment by blackHatMonkey** [comment]
Yes this patch should improve the stability a lot. In regards to test cases I thought about it a lot while writing the code however I couldn't find any solutions. Since all the functions require an adapter to be run or need manual check for finding access points for instance it is very hard to do. I'm open to suggestions if you have any. I will update my OP to include more details along a complete change log.

File: wifiphisher/interfaces.py (hunk 2 above)
Comment (sophron) [inline wifiphisher/interfaces.py]: Declaration / initialization is not required in Python. I don't see how it makes the code cleaner. :)

File: wifiphisher/interfaces.py (hunk 3 above)
Comment (sophron) [inline wifiphisher/interfaces.py]: Yes, that's pretty much  it.

File: wifiphisher/interfaces.py (hunk 4 above)
Comment (sophron) [inline wifiphisher/interfaces.py]: No, the idea of `constants.py` is to define all constants (not only those required by the main module). 

The reason is not only reusability. An advanced user that wants to configure the software should be able to open `constants.py` and mess with it, instead of trying to figure out which file to change. In this scenario, a user may be using a different version of `iw` tool on his box where the support of monitor mode is represented differently. He can now simply edit the `constants.py` file instead of opening/looking at other files.

Imagine `constants.py` as parameters for advanced use cases.

File: wifiphisher/pywifiphisher.py (hunk 8 above)
Comment (sophron) [inline wifiphisher/pywifiphisher.py]: Just the 'HINT'.

**Comment by sophron** [comment]
The best way is to probably <a href="https://docs.python.org/3/library/unittest.mock.html">mock</a> `exec_cmd` method. For example, you can test `get_compatibility` by mocking `exec_cmd` and return different outputs of `iw` from it. Different outputs should resemble different test case scenarios. While these outputs will be hard-coded in each test scenario, they should be taken from true runs of `iw`.

**Comment by sophron** [comment]
For the CHANGELOG, it's better if you provide what this PR is adding compared to what we currently have.

**Comment by sophron** [comment]
Furthermore, I believe it's better if we follow a Object Oriented approach, just like we did with `phishingpage`.

**Comment by blackHatMonkey** [comment]
I originally wrote it fallowing an object oriented approach but I didn't see any added value for making it into a class. Did you mean to put everything in one class or did you have something else in mind?

**Comment by sophron** [comment]
We need an `InterfaceManager` class, just like we have the `TemplateManager`. All the logic regarding automatically selecting an interface or a single retrieval of an interface should be contained there. 

I would also recommend a `NetworkInterface` class. This will represent all interfaces of the system, either physical or virtual. Fields of this class should define the attributes of the interface.

The `InterfaceManager` should hold a data structure (probably a dictionary) with all stored `NetworkInterface` objects. 

In general, the logic would be similar with `phishingpage` where:

* `TemplateManager` -> `InterfaceManager` 
* `PhishingTemplate` -> `NetworkInterface`

**Comment by abhiinos111abhi** [comment]
are u working on a problem with the interface ?? will this patch solve my problem?? when i use wifiphisher it use my laptop adapter to make the ap and my alfa AWUS036H use as jamming device....but my alfa device has good range so i want that ap will create with alfa device. if i use 
                                                 ./wifiphisher -aI wlan1 -jI wlan0
then it shows 

[+] Driver initialization failed! (hostapd error)
[+] Try a different wireless interface using -aI option.
wlan0 is laptop adapter
wlan1 is alfa

**Comment by blackHatMonkey** [comment]
@abhiinos111abhi 
The problem in you case is that ```AWUS036H``` doesn't support ```AP``` mode and hence why it fails. You are more than welcome to try this patch and see if it fixes you issue.
If you are going to try this patch this is how you get it.
1. Download the files from my [repository](https://github.com/blackHatMonkey/wifiphisher.git)
2. Switch to ```interface-module```
while in the ```wifiphisher``` folder
```bash
git checkout interface-module
```
3. Run the program

**Comment by abhiinos111abhi** [comment]
bad news but thanks for replying and i'll be try if the patch fixes

**Comment by sophron** [comment]
@abhiinos111abhi, @blackHatMonkey I had to delete your messages. This area is *only* for discussion regarding the development of this PR. Please don't post any more irrelevant things here.

**Comment by blackHatMonkey** [comment]
I just pushed out the changes. Let me know when you review them.:smile:

**Comment by sophron** [comment]
Thanks. That was a good PR.
